File: day:53/isListPalindrome.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        # base case
        if(head is None or head.next is None): return True;
    
        stack = [];
        current = head;
    
        # First parse
        while(current != None):
            stack.append(current.val);
            current = current.next;
        # Second parse
        current = head;
    
        while(current != None):
            if(stack.pop() != current.val):
                return False;
            current = current.next;
    
        return True;

File: day:53/isListPalindrome.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    /**
        idea: Push the nodes to a stack. Traverse again the list and compare against the each first element of  the stack. 
        If matches, keep on POPPING off the stack. If traversal done with all elements pop-off, true else false.
    **/ 
    // base case
    if(head === null || head.next === null) return true;
    
    let stack = [];
    let current = head;
    
    // First parse
    while(current !== null) {
        stack.push(current.val);
        current = current.next;
    }
    // Second parse
    current = head;
    
    while(current !== null) {
        if(stack.pop() !== current.val){
            return false;
        }
        current = current.next;
    }
    
    return true;
};
